src/MyParcelComApi.php

```diff
@@ -439,42 +444,59 @@ public function updateShipment(ShipmentInterface $shipment)
     }
 
     /**
-     * Determine what service contract to use for given shipment and update the
-     * shipment.
+     * Determine which service and contract to use for given shipment
+     * and update the shipment.
      *
      * @param ShipmentInterface $shipment
      * @return $this
      */
-    protected function determineContract(ShipmentInterface $shipment)
+    protected function determineServiceAndContract(ShipmentInterface $shipment)
     {
-        $selector = new ContractSelector();
         $calculator = new PriceCalculator();
 
-        $allServices = [];
-        $collection = $this->getServices($shipment);
-        for ($offset = 0; $offset < $collection->count(); $offset += 30) {
-            $allServices = array_merge($allServices, $collection->offset($offset)->get());
+        if ($shipment->getService() !== null) {
+            $services = [$shipment->getService()];
+        } else {
+            $services = [];
+            $collection = $this->getServices($shipment);
+            for ($offset = 0; $offset < $collection->count(); $offset += 30) {
```

Comment:
Using `30` here seems a bit magical. It would be a good idea to set the `limit` to `30` first. Then if we ever increase the default limit, it won't suddenly break.